pkg/azure/rule_utils.go

```diff
@@ -0,0 +1,189 @@
+	err := setupMaps(reservedPrioritiesInbound, reservedPrioritiesOutbound, nil, nsg)
+	if err != nil {
+		utils.Log.Printf("An error occured during setup: %+v", err)
+		return false, err
+	}
+
+	// For Inbound Rules
+	priority, err := validateSecurityRulesConform(reservedPrioritiesInbound)
+	if err != nil {
+		if priority == -1 {
+			return false, fmt.Errorf("Non-compliant: %v", err)
+		}
+
+		_, err := setupAndCreateDenyAllRule(ctx, azureHandler, priority, inboundDirectionRule, *nsg.Name)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// For Outbound Rules
+	priority, err = validateSecurityRulesConform(reservedPrioritiesOutbound)
+	if err != nil {
+		if priority == -1 {
+			return false, fmt.Errorf("Non-compliant: %v", err)
+		}
+
+		_, err := setupAndCreateDenyAllRule(ctx, azureHandler, priority, outboundDirectionRule, *nsg.Name)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// Checks that the NSG rules in a particular direction are conformant as per the description of (func) CheckSecurityRulesCompliance
+//
+// Returns the priority number of the deny all rule if it exists (with no error)
+// Or the priority number of the deny all rule to be created if the rules are non-conformant (with an error)
+// Or a priority number of -1 (with an error) to represent an invalid rule order.
+func validateSecurityRulesConform(reservedPriorities map[int32]*armnetwork.SecurityRule) (int32, error) {
+	var lowestRule *armnetwork.SecurityRule
+	lowestDenyPriorityNum := int32(maxPriority)
+	highestAllowPriorityNum := int32(minPriority)
+
+	for priority, rule := range reservedPriorities {
+		access := *rule.Properties.Access
+		if (access == armnetwork.SecurityRuleAccessAllow) && (priority > highestAllowPriorityNum) {
+			highestAllowPriorityNum = priority
+		}
+
+		if (access == armnetwork.SecurityRuleAccessDeny) && (priority <= lowestDenyPriorityNum) {
+			lowestRule = rule
+			lowestDenyPriorityNum = priority
+		}
+	}
+
+	// An allow rule's priority number is higher than a deny rule's priority number
+	if highestAllowPriorityNum > lowestDenyPriorityNum {
+		return -1, fmt.Errorf("Allow Rule with lower priority(%d) than Deny Rule(%d)", highestAllowPriorityNum, lowestDenyPriorityNum)
+	}
+
+	// No deny rule exists in the NSG, so non-conformant
+	if lowestRule == nil {
+		if reservedPriorities[maxPriority] != nil {
+			// The max priority number should not be associated to an allow rule
+			return -1, fmt.Errorf("Allow Rule at lowest priority(%d). Must be a deny all rule", maxPriority)
+		}
+
+		return maxPriority, fmt.Errorf("No deny rule present")
+	}
+
+	// If not a deny all rule, return priority to create a deny all rule
+	if !isDenyAllRule(lowestRule) {
+		lastPriority := getPriority(reservedPriorities, minPriority, maxPriority, false)
+		return lastPriority, fmt.Errorf("Deny Rule not at lowest priority(%d) is not a Deny all rule", lowestDenyPriorityNum)
+	}
+
+	return lowestDenyPriorityNum, nil
```

Comment:
Overall, the only reason we would need to check for a deny-all is if they already have one that is higher priority (lower value) than the default azure allow vnet rules. In that case, we can use that rule as our deny-all.
Otherwise, the requirement for deny rules (not specifically deny-all) is that there shouldn't be any.

So to summarize the requirements:
* No deny rules that are not deny-all
* The only acceptable deny-all rules are (1) the lowest priority in the paraglider range (100-4096) or (2) just above the default Azure rules (<65500)

We will override (1) and the default azure rules directly above it that allow things like the local vnet traffic. And for (2), we can use it as our inserted deny-all if we want. Or we can just say that (2) is not allowed (to make things simple) and then insert our deny-all for everything. 

So for (2), let's do the simple option: it doesn't cause a failure, but we still override it with a deny-all at 4096.